functions/src/modules/CR.functions.ts

```diff
@@ -0,0 +1,85 @@
+import * as functions from 'firebase-functions';
+import * as helpers from '../helpers';
+import GoogleSheets from '../services/GoogleSheets';
+import * as _ from "lodash";
+
+export const copySubmissionsToProcessing = functions
+  .pubsub
+  .topic('daily-tick')
+  .onPublish(async (message) => {
+    const onlineSheet = await
+      (new GoogleSheets(functions
+        .config()
+        .cr
+        .submissions
+        .spreadsheet
+        .id)
+      ).useSheet('Online');
+    const onlineSheetRows = await onlineSheet.getRows();
+
+    const destSpreadsheet = new GoogleSheets(functions.config().cr.processing.spreadsheet.id);
+    const destSheets = {};
+
+    const sheetNames = await destSpreadsheet.getSheetNames();
+
+    // used to hold `Submission subserials` of each sheet at the dest spreadsheet
+    const submissionSerials = {};
+
+    // Used to hold `Audio file names` of each sheet, to help look for the right place
+    // to insert a new submission
+    const audioFileNames = {};
+
+    // Holds the rows read from the `Online` sheet
+    // grouped by the sheet (list) name
+    const valuesToAppend = {};
+
+    const insertions = {};
+
+    for (const row of onlineSheetRows) {
+      const list = helpers.extractListFromFilename(row['Audio File Name']);
+
+      if (!destSheets[list]) {
```

Comment:
Let's make it systematic. When you meet a first file from some list, initialize everything at once: `destSheets`, `submissionSerials`, `audioFileNames`, `insertions`, `valuesToAppend`.

Moreover, let's use [Maps](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) instead of a bunch of plain objects. Here is my proposal:

```
const destination = new Map();
…
if (!destination.has(list)) {
  const sheet = await destSpreadsheet.useSheet(list);
  destination.set(list, {
    sheet,
    submissionSerials: await sheet.getColumn('Submission Serial'),
    audioFileNames: await sheet.getColumn('Audio File Name'),
    insertions: new Map(),
    rows: [] // This is in replace to valuesToAppend
  });
```

Your life will become easier, and the code will become more elegant.

Reply:
Thanks, that's much cleaner than what I did